pkg/controller/registry/grpc/source.go

```diff
@@ -100,10 +105,71 @@ func (s *SourceStore) Get(key registry.CatalogKey) *SourceConn {
 	return &source
 }
 
+func grpcProxyURL(addr string) (*url.URL, error) {
+	// Handle ip addresses
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	url, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hardcode fields required for proxy resolution
+	url.Host = addr
+	url.Scheme = "http"
+
+	// Override HTTPS_PROXY and HTTP_PROXY with GRPC_PROXY
+	proxyConfig := &httpproxy.Config{
+		HTTPProxy:  getGRPCProxyEnv(),
+		HTTPSProxy: getGRPCProxyEnv(),
+		NoProxy:    getEnvAny("NO_PROXY", "no_proxy"),
+		CGI:        os.Getenv("REQUEST_METHOD") != "",
+	}
+
+	// Check if a proxy should be used based on environment variables
+	return proxyConfig.ProxyFunc()(url)
+}
+
+func getGRPCProxyEnv() string {
+	return getEnvAny("GRPC_PROXY", "grpc_proxy")
+}
+
+func getEnvAny(names ...string) string {
+	for _, n := range names {
+		if val := os.Getenv(n); val != "" {
+			return val
+		}
+	}
+	return ""
+}
+
+func grpcConnection(address string) (*grpc.ClientConn, error) {
+	dialOptions := []grpc.DialOption{grpc.WithInsecure()}
+	proxyURL, err := grpcProxyURL(address)
+	if err != nil {
+		return nil, err
+	}
+
+	if proxyURL != nil {
+		dialOptions = append(dialOptions, grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			dialer, err := proxy.FromURL(proxyURL, &net.Dialer{})
+			if err != nil {
+				return nil, err
+			}
+			return dialer.Dial("tcp", addr)
+		}))
+	}
+
+	return grpc.Dial(address, dialOptions...)
+}
+
 func (s *SourceStore) Add(key registry.CatalogKey, address string) (*SourceConn, error) {
 	_ = s.Remove(key)
 
-	conn, err := grpc.Dial(address, grpc.WithInsecure())
+	conn, err := grpcConnection(address)
```

Comment:
Can we abstract the dial logic away? maybe take a `Dailer` as a field of `SourceStore`?

Reply:
If we pursue this option wouldn't the connection to the proxy need to remain open indefinitely?